### ISSUE TITLE
Validation errors applied to ListBlocks inside StructBlocks don't get rendered.

### ISSUE THREAD

**Issue by coredumperror** [wagtail/wagtail#5300]
I've got a custom `clean()` method on my StructBlock subclass, `TestBlock`. `TestBlock` contains two child blocks: `text_field` (a CharBlock) and `list_field` (a ListBlock of CharBlocks). My custom `clean()` method ensures that the user provides at least three elements to `list_field`, otherwise it throws a `ValidationError`. Here's the code:

    class TestBlock(blocks.StructBlock):

        text_field = blocks.CharBlock(
            label='Text',
            max_length=255
        )
        list_field = blocks.ListBlock(
            blocks.CharBlock(max_length=255),
            default=[],
        )

        class Meta:
            label = 'Test Block'
            template = 'app/blocks/TestBlock.html'
            form_classname = 'test-block struct-block'

        def clean(self, value):
            result = super().clean(value)
            errors = {}
            list_field_count = len(value['list_field'])
            if list_field_count < 3:
                errors['list_field'] = ErrorList(['You must provide at least three.'])
            if errors:
                # The message here is arbitrary - StructBlock.render_form will suppress it
                # and delegate the errors contained in the 'params' dict to the child blocks
                raise ValidationError('TableBlock received invalid data', params=errors)
            return result

The problem is that while the validator will successfully mark the form with an error if the user provides less than 3 elements for `list_field`, the form will not render the "You must provide at least three" message anywhere. Here's a screenshot of what I mean:
<img width="1595" alt="Screen Shot 2019-05-15 at 3 34 56 PM" src="https://user-images.githubusercontent.com/413571/57814162-a2c8a880-7727-11e9-9b43-f23d5f2b57d5.png">

However, by changing `errors['list_field']` to `errors['text_field']` in the `clean()` method, you get this:
<img width="1606" alt="Screen Shot 2019-05-15 at 3 40 39 PM" src="https://user-images.githubusercontent.com/413571/57814272-07840300-7728-11e9-9c08-03eb34a0c3c7.png">

So the error rendering is functioning properly on CharBlock. It's just not working on ListBlock, for some reason.

**Comment by komailhaider10p** [comment]
Any update on this issue? I am facing same issue.

**Comment by loicteixeira** [comment]
Looking at the `ListBlock` source code, I reckon `errors` (in your code snippet) should be a list of `ErrorList` (instead of a simple `list`) and should have the same length as the number of elements (using `None` when there are no errors): 

https://github.com/wagtail/wagtail/blob/0507ec5377c63cf354d5740598a9d3babc289950/wagtail/core/blocks/list_block.py#L116-L130

I won't be tagging this as bug for now as it looks like a misuse of the API. Maybe there would be some enhancement or at least some documentation but probably not a bug.

**Comment by komailhaider10p** [comment]
Thanks Loic for your response.

I am a bit new to django/wagtail. Can you please share what changes do I 
need to make in my code snippet to make it working. Here it is given below:

I am getting below error while running this:

Exception Type: 	AttributeError
Exception Value: 	

'list' object has no attribute 'get'

Exception Location: 
/usr/local/lib/python3.6/site-packages/wagtail/core/blocks/struct_block.py 
in <listcomp>, line 94


from django.core.exceptions import ValidationError
from django.forms.utils import ErrorList
classTestBlock(blocks.StructBlock):
     text_field = blocks.CharBlock(
label='Text',
max_length=255
     )
     list_field = blocks.ListBlock(
         blocks.CharBlock(max_length=255),
default=[],
     )
classMeta:
         label ='Test Block'
         template ='app/blocks/TestBlock.html'
         form_classname ='test-block struct-block'
defclean(self, value):
         result = []
         errors = []
         list_field_count =len(value['list_field'])
if list_field_count <3:
for child_val in value:
                 errors.append(ErrorList(["Error"]))
else:
             errors.append(None)
ifany(errors):
# The message here is arbitrary - outputting error messages is delegated to the child blocks,
# which only involves the 'params' list
raise ValidationError('Validation error in ListBlock', params=errors)
return result

**Comment by loicteixeira** [comment]
Hi @komailhaider10p. I would like to keep this issue about finding whether the initial report should be classified as a bug.

At this point, [Stack Overflow](https://stackoverflow.com/questions/tagged/wagtail) (with the `Wagtail` tag) or [Slack](https://docs.wagtail.io/en/stable/support.html#slack) would be a better place to find a solution to your specific implementation.

My bad for not flagging this immediately and proposing a potential fix.

**Comment by bmoe872** [comment]
@loicteixeira I'm running into this same problem, and I do think this is a bug. 

With the explanation that was given, I don't see how it's not a bug. The code snippet your provided is the `clean` method for the ListBlock, but the original problem is the override of the `clean` method on the StructBlock specifically, wherein overriding this should apply the appropriate error to the list block itself, and not "Within" the list block. 

Further, even if the ListBlock has a different definition for Clean, then it should behave the same as other fields within a StructBlock no? i.e. the clean definition on ListBlock should not interfere with the StructBlock's clean method.

**Comment by gasman** [comment]
I agree with @loicteixeira's assessment. Validation error handling on StreamField blocks works on the contract that whenever a block's `clean` method throws an exception, that same exception object will ultimately be passed back to that block's `render_form` method to be rendered back to the user.

Since the top-level StreamField's `clean` method can only throw one exception back to Django's forms framework - even if there are multiple errors within it - any block that acts as a container for child blocks (such as StructBlock and ListBlock) is responsible for bundling up the exceptions thrown by its children into a single `ValidationError`, and then unbundling them again in `render_form` to pass each of those exceptions back to the relevant child's `render_form` method.

It's important to note that this "bundling" mechanism is private to each container block type, and any `ValidationError` objects emerging from it should be treated as a "black box" - they are liable to contain a specially-crafted `params` dict that is only understood by that one block.

The line `errors['list_field'] = ErrorList(['You must provide at least three.'])` is breaking this contract - it's constructing a `ValidationError` that will ultimately be pushed to `ListBlock.render_form` but doesn't follow `ListBlock`'s internal format for `ValidationError`s.

The piece that's really missing here is that `ListBlock` currently has no concept of "non field errors" - it assumes that any validation errors are the responsibility of one of its child blocks. Once you implement that, then you're 90% of the way towards supporting `min_num` / `max_num` arguments on `ListBlock`, which I suspect is the feature you're _really_ looking for here :-) And, indeed, one of the things #3914 implements is [a ValidationError subclass with provision for non-block errors](https://github.com/wagtail/wagtail/pull/3914/files#diff-dba47bd18dcf91593c528c6c3e121a15R20), to match [the one that already exists for StreamBlock](https://github.com/wagtail/wagtail/blob/d37d5c1c60509e34c09add8ade94a07a87c3a867/wagtail/core/blocks/stream_block.py#L22).

(I'm half inclined to close this as a duplicate of #2379 - but on the other hand #3914 is largely blocked by the Javascript code, which would be part of a complete `min_num` / `max_num` solution but out of the scope of what's being discussed here - so I think we can consider the server-side validation to be a useful stepping stone. In other words: adding server-side validation for `min_num` / `max_num` on ListBlock will close this ticket, and adding the corresponding constraints on the client-side UI will close #2379.)
